ctru-rs/src/os.rs

```diff
@@ -0,0 +1,154 @@
+// define enums for firmware type (NATIVE_FIRM, SAFE_FIRM etc.) as well as
+// application memory layout. Leaving those as future enhancements for now
+
+/// A region of memory. Most applications will only use [`Application`](MemRegion::Application)
+/// memory, but the other types can be used to query memory usage information.
+/// See <https://www.3dbrew.org/wiki/Memory_layout#FCRAM_memory-regions_layout>
+/// for more details on the different types of memory.
+///
+/// # Example
+/// ```
+/// # let _runner = test_runner::GdbRunner::default();
+/// let all_memory = ctru::os::MemRegion::All;
+///
+/// assert!(all_memory.size() > 0);
+/// assert!(all_memory.used() > 0);
+/// assert!(all_memory.free() > 0);
+/// ```
+#[derive(Clone, Copy, Debug)]
+#[non_exhaustive]
+#[repr(u32)]
+pub enum MemRegion {
+    /// All memory regions.
+    All = ctru_sys::MEMREGION_ALL,
+    /// APPLICATION memory.
+    Application = ctru_sys::MEMREGION_APPLICATION,
+    /// SYSTEM memory.
+    System = ctru_sys::MEMREGION_SYSTEM,
+    /// BASE memory.
+    Base = ctru_sys::MEMREGION_BASE,
+}
+
+impl MemRegion {
+    /// Get the total size of this memory region, in bytes.
+    pub fn size(&self) -> usize {
+        unsafe { ctru_sys::osGetMemRegionSize(*self as u32) }
+            .try_into()
+            .unwrap()
+    }
+
+    /// Get the number of bytes used within this memory region.
+    pub fn used(&self) -> usize {
+        unsafe { ctru_sys::osGetMemRegionUsed(*self as u32) }
+            .try_into()
+            .unwrap()
+    }
+
+    /// Get the number of bytes free within this memory region.
+    pub fn free(&self) -> usize {
+        unsafe { ctru_sys::osGetMemRegionFree(*self as u32) }
+            .try_into()
+            .unwrap()
+    }
+}
+
+/// WiFi signal strength. This enum's `u8` representation corresponds with
+/// the number of bars displayed in the Home menu.
+///
+/// # Example
+///
+/// ```
+/// # let _runner = test_runner::GdbRunner::default();
+/// let strength = ctru::os::WifiStrength::current();
+/// assert!((strength as u8) < 4);
+/// ```
+#[derive(Clone, Copy, Debug)]
+#[non_exhaustive]
+#[repr(u8)]
+pub enum WifiStrength {
+    /// This may indicate a very poor signal quality even worse than `Bad`,
+    /// or that no network is connected at all.
+    Disconnected = 0,
+    /// Poor signal strength.
+    Bad = 1,
+    /// Medium signal strength.
+    Decent = 2,
+    /// Good signal strength.
+    Good = 3,
+}
+
+impl WifiStrength {
+    /// Get the current WiFi signal strength.
+    pub fn current() -> Self {
+        match unsafe { ctru_sys::osGetWifiStrength() } {
+            0 => Self::Disconnected,
+            1 => Self::Bad,
+            2 => Self::Decent,
+            3 => Self::Good,
+            other => panic!("Got unexpected WiFi strength value {other}"),
+        }
+    }
+}
+
+/// Get the current value of the stereoscopic 3D slider on a scale from 0.0­–­1.0.
+pub fn current_3d_slider_state() -> f32 {
```

Comment:
I intended to add this in tandem with some other changes to the `hid` service I'm working with. Sure, it's not required to use `Hid` in this circumstance, but it makes much more sense (since `hid` handles hardware input by definition) and since I'm also working on using the volume slider (which instead *requires* the `hid` service).

I don't like the redundancy usually found within `libctru` (usually generated by some OS-level redundancy...), so I'd like to avoid having this function in two places. Do you think we should keep it here or move it to hid?

Reply:
Well, from what I can tell reading `libctru` code, the main difference is:

* functions defined in `os.h` are a simple pointer read of a well-known address `((osSharedConfig_s*)OS_SHAREDCFG_VADDR)->slider_3d`

* HID functions use `svcSendSyncRequest` IPC, which is also why it requires the HID service to be initialized first

I put this in here to match `libctru`'s structure, and after looking more carefully at the implementation I think it probably makes sense to keep it that way. Since these functions don't require an initialized HID service to be used, I'm not sure it makes sense to add them to our `Hid` implementation either.